src/NHibernate/Linq/GroupResultOperatorExtensions.cs

```diff
@@ -19,12 +19,31 @@ private static IEnumerable<Expression> ExtractKeyExpressions(this Expression exp
 		{
 			// Recursively extract key expressions from nested initializers 
 			// --> new object[] { ((object)new object[] { x.A, x.B }), x.C }
-			// --> x.A, x.B, x.C
-			if (expr is NewExpression)
-				return (expr as NewExpression).Arguments.SelectMany(ExtractKeyExpressions);
-			if (expr is NewArrayExpression)
-				return (expr as NewArrayExpression).Expressions.SelectMany(ExtractKeyExpressions);
+			// --> new List<string> { x.A, x.B }
+			// --> new CustomType(x.A, x.B) { C = x.C }
```

Comment:
Not sure if this is correct. There are examples when this assumption is not working.

Reply:
It would be incorrect for Linq to Objects, but in our scenario the group by is done on the database, so the constructor of `CustomType2` will never be called. In our case the following expressions will produce the same result:
```C#
GroupBy(x => new List<string> { x.A, x.B, x.C })
GroupBy(x => new CustomType(x.A, x.B) { C = x.C })
GroupBy(x => new { x.A x.B, x.C })
```
The idea of this method is to collect all mapped properties inside the expression and use them as keys for grouping.

Reply:
I think this might be restricted to some known types (potentially a configurable option) and some heuristics (like looking for [CompilerGeneratedAttribute]).

Reply:
> so the constructor of CustomType2 will never be called. 

What about case when group key is selected?
```C#
GroupBy(x => new CustomType(x.A, x.B) { C = x.C }).Select(x => x.Key);
```
In any case I think it's better keep the code simpler and allow it without any added heuristics (and I've seen multiple requests specifically to support ctors for user DTOs for whatever legacy reasons).

> There are examples when this assumption is not working.

Well it applies for all cases with user types - user can override equality methods or getters with some strange logic and that's it.

Reply:
> There are examples when this assumption is not working.

This concern already holds true for current master state. I do not think we should consider this as added or worsen by this PR. So I think it should be addressed in another PR, if it actually needs to be addressed.

Reply:
> I think this might be restricted to some known types (potentially a configurable option) and some heuristics (like looking for [CompilerGeneratedAttribute])

What would happen when the restriction is not met? For example the following query:
```C#
_session.Query<X>().GroupBy(x => new CustomType(x.A, x.B) { C = x.C }).ToList();
```
if the restriction is not met we could either throw an exception or execute `_session.Query<X>()` and do the rest on the client, which I don't thnik we should do. 

> like looking for [CompilerGeneratedAttribute]

In OData, the [wrappers that they use](https://github.com/OData/WebApi/blob/279f60dabac9c6de2bb8f560886ffa499940bc89/src/Microsoft.AspNet.OData.Shared/Query/Expressions/DynamicTypeWrapper.cs#L112) have only `JsonConverter` attribute which I don't think we should check. The only restriction that we may add, is to check whether the type overriders `Equals` and `GetHashCode` methods and throw when it does. But even that restriction does not make much sense as grouping in Linq To Objects by a type that does not override those methods won't group any items together, where in our case it may group from zero to all items together. In my opinion we should by default allow all types and optionally having a configurable option that would throw when the given restriction is not met.

> What about case when group by key is selected?

In that case the constructor will be called for each group. The following query produces the same result:
```C#
GroupBy(x => new {x.A, x.B, x.C }).Select(x => new CustomType(x.Key.A, x.Key.B) { C = x.Key.C }); 
```
When the user wants to return all groups as a list of `CustomType` then IMO your query is better as it is shorter to write.

Reply:
> to check whether the type overrides Equals and GetHashCode methods and throw when it **does**

I don't get it.  Wouldn't it throw for all anonymous types? Did you mean "and throw when it **doesn't"**? If I got it right the main concern was about difference between Linq to Objects and NHibernate results. And from this context user type must override equality members to have the same results.
But there is so many ways that user can break the difference in results so I wouldn't implement any additional restrictions at all.

Reply:
> I don't get it. Why? 

Whether we would want to prevent the user from thinking that the `Equals` and `GetHashCode` would be used for grouping.

> Did you mean and throw when it **doesn't"**?

No, I ment when it does override those methods it would throw.

> But there is so many ways that user can break the difference in results so I wouldn't implement any additional restrictions at all.

True, that is why I also agree that we should not add any restriction by default.

Also, I have to correct myself as in the following case the constructor will also be called:
```C#
_session.Query<X>().GroupBy(x => new CustomType(x.A, x.B) { C = x.C }).Select(o => new { o.Key.A, o.Key.B, o.Key.C }).ToList();
```
beacuse there is no explicit link in the expression between `o.Key.A\o.Key.B` and `x.A\x.B`, due to the fact that `x.A` and `x.B` are passed to the constructor.
Which means that it will work correctly even when `x.A` is set in the constructor to property `CustomType.B`.